From c5b0cf161d14af6f2ff4de3cf7ce201e6070ef88 Mon Sep 17 00:00:00 2001
From: Pankaj Bharadiya <pankaj.laxminarayan.bharadiya@intel.com>
Date: Fri, 23 Feb 2018 16:50:12 +0530
Subject: [PATCH 1468/2367] Soundwire: Fix error handling in
 sdw_get_stream_tag()

In sdw_get_stream_tag(), the sdw_rt pointer is dereferenced without
checking for the validity.
Check for the pointer validity before dereferencing it. Also unlock
core_lock mutex in case of error.

Change-Id: I4241f97965186015d79320537bd86aacc7fa1d60
Acked-by: Vaibhav Agarwal <vaibhav.agarwal@intel.com>
Signed-off-by: Pankaj Bharadiya <pankaj.laxminarayan.bharadiya@intel.com>
---
 drivers/sdw/sdw.c | 12 ++++++------
 1 file changed, 6 insertions(+), 6 deletions(-)

diff --git a/drivers/sdw/sdw.c b/drivers/sdw/sdw.c
index ee09aed..30e75cc 100644
--- a/drivers/sdw/sdw.c
+++ b/drivers/sdw/sdw.c
@@ -2229,20 +2229,20 @@ static int sdw_get_stream_tag(char *key, int *stream_tag)
 key_check_not_required:
 	for (i = 0; i < SDW_NUM_STREAM_TAGS; i++) {
 		if (!stream_tags[i].ref_count) {
-			stream_tags[i].ref_count++;
 			*stream_tag = stream_tags[i].stream_tag;
 			mutex_init(&stream_tags[i].stream_lock);
 			sdw_rt = kzalloc(sizeof(struct sdw_runtime),
 					GFP_KERNEL);
+			if (!sdw_rt) {
+				ret = -ENOMEM;
+				mutex_unlock(&sdw_core.core_lock);
+				goto out;
+			}
+			stream_tags[i].ref_count++;
 			INIT_LIST_HEAD(&sdw_rt->slv_rt_list);
 			INIT_LIST_HEAD(&sdw_rt->mstr_rt_list);
 			sdw_rt->stream_state = SDW_STATE_INIT_STREAM_TAG;
 			stream_tags[i].sdw_rt = sdw_rt;
-			if (!stream_tags[i].sdw_rt) {
-				stream_tags[i].ref_count--;
-				ret = -ENOMEM;
-				goto out;
-			}
 			if (key)
 				strlcpy(stream_tags[i].key, key,
 					SDW_MAX_STREAM_TAG_KEY_SIZE);
-- 
2.7.4

